Introduce `QuerySetAny` type for QuerySet isinstance checks (#1199)
TypedDjango/django-stubs: README.md

func(MyModel.objects.annotate(bar=Value("")).get(id=1))  # Error
```

### How do I check if something is an instance of QuerySet in runtime?

A limitation of making `QuerySet` generic is that you can not use
it for `isinstance` checks.

```python
from django.db.models.query import QuerySet

> def foo(obj: object) -> None:
>     isinstance(obj, QuerySet) # Error: Parameterized generics cannot be used with class or instance checks
> ```
> 
> To get around with this issue without making `QuerySet` non-generic,
> Django-stubs provides `django_stubs_ext.QuerySetAny`, a non-generic
> variant of `QuerySet` dedicated for runtime type checking:
> 
> ```python
> from django_stubs_ext import QuerySetAny
> 
> def foo(obj: object) -> None:
>     isinstance(obj, QuerySetAny)  # OK

Reviewer: Can we make the examples a little bit more realistic?

```suggestion
def foo(obj: object) -> None:
    if isinstance(obj, QuerySet): # Error: Parameterized generics cannot be used with class or instance checks
        ...
```

To get around with this issue without making `QuerySet` non-generic,
Django-stubs provides `django_stubs_ext.QuerySetAny`, a non-generic
variant of `QuerySet` dedicated for runtime type checking:

```python
from django_stubs_ext import QuerySetAny

def foo(obj: object) -> None:
    if isinstance(obj, QuerySetAny):  # OK
        ...
```

Author: Good point! Fixed.

func(MyModel.objects.annotate(bar=Value("")).get(id=1))  # Error
```

### How do I check if something is an instance of QuerySet in runtime?

A limitation of making `QuerySet` generic is that you can not use
it for `isinstance` checks.

```python
from django.db.models.query import QuerySet

def foo(obj: object) -> None:
    if isinstance(obj, QuerySet): # Error: Parameterized generics cannot be used with class or instance checks
        ...
```

To get around with this issue without making `QuerySet` non-generic,
Django-stubs provides `django_stubs_ext.QuerySetAny`, a non-generic
variant of `QuerySet` suitable for runtime type checking:

```python

> from django_stubs_ext import QuerySetAny

Reviewer 2: Question for other PRs: do we have other types to take care of?

func(MyModel.objects.annotate(bar=Value("")).get(id=1))  # Error
```

### How do I check if something is an instance of QuerySet in runtime?

A limitation of making `QuerySet` generic is that you can not use
it for `isinstance` checks.

```python
from django.db.models.query import QuerySet

def foo(obj: object) -> None:
    isinstance(obj, QuerySet) # Error: Parameterized generics cannot be used with class or instance checks
```

To get around with this issue without making `QuerySet` non-generic,
Django-stubs provides `django_stubs_ext.QuerySetAlias`, a non-generic
variant of `QuerySet` dedicated for runtime type checking:

```python

> from django_stubs_ext import QuerySetAlias

Reviewer 2: Do others have any name ideas for this?

I fear that `Alias` is too generic. It does not show the purpose of this thing.

Maybe `QuerySetInstance`?

Reviewer: `QuerySetInstance` 👍

Reviewer 3: Alternatively `QuerySetAny` as in "QuerySet of any model"?

Author: `QuerySetAny` sounds better, since this type is essentially `_QuerySet[Any, Any]`.
